File: ROS/ROS_21_Tutorials/catkin_ws/src/learning_topic/src/personSubscriber.cpp
/**
 * @file    personSubscriber.cpp.c
 * @brief   Subscriber
 * Details: 
 * @author  jackzhang
 * @email   zhangmengwei1996@outlook.com
 * @version 1.0.0
 * @date    3/10/21
 * @license GNU General Public License (GPL)
 */
#include <ros/ros.h>
#include "learning_topic/Person.h"

void personCallback(const learning_topic::Person::ConstPtr &personMsg)
{
	ROS_INFO("Got Person Info: name: %s, age: %d, sex: %d",
			 personMsg->name.c_str(), personMsg->age, personMsg->sex);
}

int main(int argc, char **argv)
{
	// 初始化ROS节点
	ros::init(argc, argv, "personSubscriber");

	ros::NodeHandle nodeHandle;

	// 创建一个节点，订阅的topic为/person_info，队列长度10, 回调函数是personCallback
	ros::Subscriber personSubscriber = nodeHandle.subscribe("/person_info", 10, personCallback);

	// 循环等待回调函数
	ros::spin();

	return 0;
}

File: ROS/ROS_21_Tutorials/catkin_ws/src/learning_topic/scripts/person_subscriber.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import rospy
from learning_topic.msg import Person


def person_callback(person_msg):
    rospy.loginfo("Got Person Info: name: %s, age: %d, sex: %d", person_msg.name, person_msg.age, person_msg.sex)


def person_subscriber():
    # 初始化ROS节点
    rospy.init_node("person_subscriber", anonymous=True)

    # 创建Subscriber， topic名称为/person_info， 数据类型为Person， 回调函数为person_callback
    rospy.Subscriber("/person_info", Person, callback=person_callback)

    rospy.spin()


if __name__ == "__main__":
    person_subscriber()
